File: auth/examples/login_example.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

// LoginRequest represents the login credentials
type LoginRequest struct {
	Email    string `json:"email"`
	Password string `json:"password"`
}

// TokenResponse represents the login response
type TokenResponse struct {
	Status  string `json:"status"`
	Message string `json:"message"`
	Data    struct {
		User struct {
			ID    uint   `json:"id"`
			Name  string `json:"name"`
			Email string `json:"email"`
			Role  string `json:"role"`
		} `json:"user"`
		Tokens struct {
			AccessToken  string `json:"access_token"`
			RefreshToken string `json:"refresh_token"`
			ExpiresIn    int64  `json:"expires_in"`
		} `json:"tokens"`
	} `json:"data"`
}

// ProfileResponse represents the profile response
type ProfileResponse struct {
	Status string `json:"status"`
	Data   struct {
		User struct {
			ID            uint    `json:"id"`
			Name          string  `json:"name"`
			Email         string  `json:"email"`
			Age           int64   `json:"age"`
			Gender        string  `json:"gender"`
			Weight        float64 `json:"weight"`
			Height        float64 `json:"height"`
			Goal          string  `json:"goal"`
			ActivityLevel string  `json:"activity_level"`
			Role          string  `json:"role"`
		} `json:"user"`
	} `json:"data"`
}

// Login authenticates a user and returns a JWT token
func Login(email, password string) (*TokenResponse, error) {
	// Create login request
	loginReq := LoginRequest{
		Email:    email,
		Password: password,
	}

	// Convert to JSON
	jsonData, err := json.Marshal(loginReq)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal login request: %w", err)
	}

	// Create HTTP request
	req, err := http.NewRequest("POST", "http://localhost:8080/login", bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	// Execute request
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	// Read response body
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	// Check for errors
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("login failed with status %d: %s", resp.StatusCode, string(body))
	}

	// Parse response
	var tokenResp TokenResponse
	if err := json.Unmarshal(body, &tokenResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &tokenResp, nil
}

// FetchProfile retrieves the user profile using the access token
func FetchProfile(accessToken string) (*ProfileResponse, error) {
	// Create HTTP request - updated path to avoid conflicts
	req, err := http.NewRequest("GET", "http://localhost:8080/api/auth/profile", nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	// Add authorization header
	req.Header.Set("Authorization", "Bearer "+accessToken)

	// Execute request
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	// Read response body
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	// Check for errors
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("profile fetch failed with status %d: %s", resp.StatusCode, string(body))
	}

	// Parse response
	var profileResp ProfileResponse
	if err := json.Unmarshal(body, &profileResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &profileResp, nil
}

func main() {
	// Example credentials - replace with actual values
	email := "user@example.com"
	password := "password123"

	// Login to get tokens
	fmt.Println("Logging in...")
	tokenResp, err := Login(email, password)
	if err != nil {
		fmt.Printf("Login error: %v\n", err)
		return
	}

	fmt.Printf("Login successful for user: %s (ID: %d)\n", 
		tokenResp.Data.User.Name, 
		tokenResp.Data.User.ID)
	
	fmt.Printf("Access token: %s\n", tokenResp.Data.Tokens.AccessToken[:20]+"...")
	fmt.Printf("Expires in: %d seconds\n", tokenResp.Data.Tokens.ExpiresIn)

	// Use the token to fetch the user profile
	fmt.Println("\nFetching user profile...")
	profileResp, err := FetchProfile(tokenResp.Data.Tokens.AccessToken)
	if err != nil {
		fmt.Printf("Profile fetch error: %v\n", err)
		return
	}

	// Display user profile
	user := profileResp.Data.User
	fmt.Printf("User Profile:\n")
	fmt.Printf("  Name: %s\n", user.Name)
	fmt.Printf("  Email: %s\n", user.Email)
	fmt.Printf("  Age: %d\n", user.Age)
	fmt.Printf("  Gender: %s\n", user.Gender)
	fmt.Printf("  Weight: %.1f\n", user.Weight)
	fmt.Printf("  Height: %.1f\n", user.Height)
	fmt.Printf("  Goal: %s\n", user.Goal)
	fmt.Printf("  Activity Level: %s\n", user.ActivityLevel)
} 

File: auth/examples/login_example.js
// Example of how to use the login API from a client application

/**
 * Example login function that authenticates a user and returns the JWT token
 * @param {string} email - User's email
 * @param {string} password - User's password
 * @returns {Promise<Object>} Token and user data
 */
async function loginUser(email, password) {
  try {
    const response = await fetch('http://localhost:8080/login', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        email: email,
        password: password,
      }),
    });

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.message || 'Login failed');
    }

    const data = await response.json();
    
    // Store tokens in localStorage or secure cookie
    localStorage.setItem('access_token', data.data.tokens.access_token);
    localStorage.setItem('refresh_token', data.data.tokens.refresh_token);
    
    // Return the user data and tokens
    return {
      user: data.data.user,
      tokens: data.data.tokens,
    };
  } catch (error) {
    console.error('Login error:', error);
    throw error;
  }
}

/**
 * Example of making an authenticated API request using the JWT token
 * @param {string} url - API endpoint
 * @returns {Promise<Object>} Response data
 */
async function makeAuthenticatedRequest(url) {
  try {
    // Get the token from storage
    const token = localStorage.getItem('access_token');
    
    if (!token) {
      throw new Error('No authentication token found');
    }
    
    const response = await fetch(url, {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
    });
    
    // If token is expired or invalid
    if (response.status === 401) {
      // Attempt to refresh the token
      const refreshed = await refreshToken();
      if (refreshed) {
        // Retry the request with the new token
        return makeAuthenticatedRequest(url);
      } else {
        throw new Error('Session expired, please login again');
      }
    }
    
    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.message || 'Request failed');
    }
    
    return response.json();
  } catch (error) {
    console.error('Request error:', error);
    throw error;
  }
}

/**
 * Example of refreshing an expired token
 * @returns {Promise<boolean>} Success status
 */
async function refreshToken() {
  try {
    const refreshToken = localStorage.getItem('refresh_token');
    
    if (!refreshToken) {
      return false;
    }
    
    const response = await fetch('http://localhost:8080/auth/refresh', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        refresh_token: refreshToken,
      }),
    });
    
    if (!response.ok) {
      // Clear tokens if refresh failed
      localStorage.removeItem('access_token');
      localStorage.removeItem('refresh_token');
      return false;
    }
    
    const data = await response.json();
    
    // Update tokens in storage
    localStorage.setItem('access_token', data.access_token);
    localStorage.setItem('refresh_token', data.refresh_token);
    
    return true;
  } catch (error) {
    console.error('Token refresh error:', error);
    return false;
  }
}

/**
 * Example of logging out a user
 */
function logoutUser() {
  // Clear tokens from storage
  localStorage.removeItem('access_token');
  localStorage.removeItem('refresh_token');
}

// Example usage:
// Login example
document.getElementById('login-form')?.addEventListener('submit', async (e) => {
  e.preventDefault();
  const email = document.getElementById('email').value;
  const password = document.getElementById('password').value;
  
  try {
    const result = await loginUser(email, password);
    console.log('Login successful:', result);
    
    // Update UI or redirect
    window.location.href = '/dashboard';
  } catch (error) {
    console.error('Login failed:', error);
    // Show error message to user
  }
});

// Protected API call example
async function getUserProfile() {
  try {
    // Updated path to avoid conflicts
    const profileData = await makeAuthenticatedRequest('http://localhost:8080/api/auth/profile');
    console.log('User profile:', profileData);
    
    // Update UI with profile data
    document.getElementById('user-name').textContent = profileData.data.user.name;
  } catch (error) {
    console.error('Failed to get profile:', error);
    
    // If authentication error, redirect to login
    if (error.message === 'Session expired, please login again') {
      window.location.href = '/login';
    }
  }
} 